### PR TITLE
[AutoDiff] [stdlib] Add operators for custom gradients.

### PR DESCRIPTION
**Pull request by rxwei**
For demo purposes.

### CODE REVIEW

File: test/AutoDiff/custom_derivatives.swift [hunk 1]
```diff
@@ -84,4 +84,10 @@ CustomDerivativesTests.test("Retroactive") {
   expectEqual(100, gradient(at: 3, in: functionWithRetroDeriv))
 }
 
+CustomDerivativesTests.test("ModifyGradientOfSquared") {
```
Comment (marcrasi) [inline test/AutoDiff/custom_derivatives.swift]: ModifyGradientOfSum?

File: stdlib/public/core/AutoDiff.swift [hunk 2]
```diff
@@ -170,6 +170,39 @@ public func differentiableFunction<T, U, R>(
   return original
 }
 
+public extension Differentiable {
+  @differentiable(wrt: self, vjp: _vjpWithGrad)
+  func withGradient(_ body: @escaping (inout CotangentVector) -> Void) -> Self {
```
Comment (marcrasi) [inline stdlib/public/core/AutoDiff.swift]: `withGradient(_ body: @escaping (CotangentVector) -> CotangentVector)` seems like a more natural API at first glance. Do you have a reason to do it this way?

Comment (marcrasi) [reply stdlib/public/core/AutoDiff.swift]: Discussed in person

**Review by marcrasi — APPROVED** [review]

File: stdlib/public/core/AutoDiff.swift (hunk 2 above)
Comment (rxwei) [reply stdlib/public/core/AutoDiff.swift]: I think my core rationale is that `withXXX` APIs in Swift usually return whatever the closure returns, e.g. `withUnsafePointer(_:)` and `withoutAcutallyEscaping(_:)`, so `x.withGradient { $0 + 1 }` would syntactically indicate to a Swift programmer that this is "returning the gradient plus one in the forward pass".

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**